instrumentation/http-url-connection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpurlconnection/HttpUrlConnectionInstrumentationModule.java

```diff
@@ -79,75 +81,65 @@ public HttpUrlConnectionInstrumentationModule() {
+      if (!tracer().shouldStartSpan(parentContext)) {
+        return;
       }
 
-      ContextStore<HttpURLConnection, HttpUrlState> contextStore =
-          InstrumentationContext.get(HttpURLConnection.class, HttpUrlState.class);
-      HttpUrlState state = contextStore.putIfAbsent(connection, HttpUrlState::new);
-
-      synchronized (state) {
-        if (!state.initialized) {
-          Context parentContext = currentContext();
-          if (tracer().shouldStartSpan(parentContext)) {
-            state.context = tracer().startSpan(parentContext, connection, connection);
-            if (!connected) {
-              scope = state.context.makeCurrent();
-            }
-          }
-          state.initialized = true;
-        }
+      // putting into storage for a couple of reasons:
+      // - to start an operation in connect() and end it in getInputStream()
+      // - to avoid creating new operation on multiple subsequent calls to getInputStream()
+      ContextStore<HttpURLConnection, Context> storage =
+          InstrumentationContext.get(HttpURLConnection.class, Context.class);
+      context = storage.get(connection);
+
+      if (context == null) {
+        context = tracer().startSpan(parentContext, connection);
+        storage.put(connection, context);
       }
-      return state;
+
+      scope = context.makeCurrent();
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void methodExit(
-        @Advice.Enter HttpUrlState state,
         @Advice.This HttpURLConnection connection,
         @Advice.FieldValue("responseCode") int responseCode,
         @Advice.Thrown Throwable throwable,
         @Advice.Origin("#m") String methodName,
-        @Advice.Local("otelScope") Scope scope) {
-
-      if (scope != null) {
-        scope.close();
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") CallDepth callDepth) {
+      if (callDepth.decrementAndGet() > 0) {
+        return;
       }
-      if (state == null) {
+      if (scope == null) {
         return;
       }
-      CallDepthThreadLocalMap.reset(HttpURLConnection.class);
-
-      synchronized (state) {
-        if (state.context != null && !state.finished) {
-          if (throwable != null) {
-            tracer().endExceptionally(state.context, throwable);
-            state.finished = true;
-          } else if ("getInputStream".equals(methodName)) {
-            // responseCode field is sometimes not populated.
-            // We can't call getResponseCode() due to some unwanted side-effects
-            // (e.g. breaks getOutputStream).
-            if (responseCode > 0) {
-              tracer().end(state.context, new HttpUrlResponse(connection, responseCode));
-              state.finished = true;
-            }
-          }
+      scope.close();
+      if (Span.fromContext(context).isRecording()) {
```

Comment:
We don't have similar condition in other instrumentations, right? Why do we need it here?

Reply:
`isRecording()` is the same as previous `!state.finished`. No idea why it was used here though

Reply:
It's needed here because HttpURLConnection is a messy API and so instrumenting it is messy.

But I do think of checking `isRecording` as more of an optimization for avoiding costly things when Sampling is enabled, so think the usage here is a bit confusing, so brought back `HttpUrlState.finished` to track this (and added more comments).